Article Creation Endpoint Migration and Improvements

agusbusta opened the pull request:
### Overview
This pull request migrates and enhances the article creation functionality from the News Creator folder to the Articles section with significant improvements in validation, error handling, and documentation.

### Key Changes
#### 1. Endpoint Refactoring
- Migrated from `/articles/create` to `/article`
- Improved request handling and validation
- Enhanced error response mechanisms

#### 2. Validation Enhancements
- Comprehensive input validation
- Type checking for all fields
- Length and format validations
- Duplicate article prevention

#### 3. Error Handling Improvements
- Specific error codes and messages
- Graceful error handling for different scenarios
- Standardized response format

#### 4. Image Processing
- Robust image download and processing
- S3 upload for original and processed images
- Error handling for image-related operations

### Technical Details
- Created modular services for validation and image processing
- Added comprehensive logging
- Implemented detailed Swagger/OpenAPI documentation
- Standardized response structures

### New Validation Rules
- `title`: 
 - Required
 - Minimum 5 characters
- `content`: 
 - Required
 - Minimum 20 characters
- Added type checking for all fields
- Prevent duplicate articles based on title

### Response Codes
- `201`: Successful article creation
- `400`: Validation errors
- `409`: Duplicate article
- `500`: Internal server error

daviddflix commented on app/routes/articles/articles.py:
```diff
@@ -205,5 +208,113 @@ def delete_article(article_id):
+            return jsonify(create_response(
+                error='An article with this title already exists'
+            )), 409
+        
```
To ensure robust validation for potential duplicate articles, I recommend making the title check case-insensitive. This will prevent scenarios where titles with different cases (e.g., Hello World vs. hello world) are incorrectly treated as distinct.

```python
from sqlalchemy import func

existing_article = Article.query.filter(
    func.lower(Article.title) == data['title'].lower()
).first()
```

daviddflix commented on app/routes/articles/articles.py:
```diff
@@ -205,5 +208,113 @@ def delete_article(article_id):
+        - title (str, required): Article title
+        - content (str, required): Article content
+        - analysis (str, optional): Article analysis
+        - used_keywords (list, optional): Keywords used in the article
```
We should not request the used keyword, since this is set when the article matches keywords in the whitelist and blacklist tables. you can leave this field empty when saving it.

daviddflix commented on app/routes/articles/articles.py:
```diff
@@ -205,5 +208,113 @@ def delete_article(article_id):
+    Request Body:
+        - title (str, required): Article title
+        - content (str, required): Article content
+        - analysis (str, optional): Article analysis
```
You can leave the analysis field as empty as well.

daviddflix commented on app/routes/articles/articles.py:
```diff
@@ -205,5 +208,113 @@ def delete_article(article_id):
+        - content (str, required): Article content
+        - analysis (str, optional): Article analysis
+        - used_keywords (list, optional): Keywords used in the article
+        - is_article_efficient (str, optional): Efficiency rating
```
For this field as the articles are supposed to be good, we can set this field to 'Green - {comment or empty}' the comment (is_article_efficient ) will be received but not required.

daviddflix commented on app/routes/articles/articles.py:
```diff
@@ -205,5 +208,113 @@ def delete_article(article_id):
+            content=data['content'].replace("- ", ""),
+            image=image_filename,
+            analysis=data.get('analysis', ''),
+            url="Generated By Ai-Alpha Team.",
```
Please adjust to "Generated By AI Alpha Team"

daviddflix commented on app/routes/articles/articles.py:
```diff
@@ -205,5 +208,113 @@ def delete_article(article_id):
+    """
+    Create a new article with comprehensive validation and error handling.
+    
+    Request Body:
```
Please, make sure to detail what parameters are required, this should be set here and in the documentation.

agusbusta commented:
# List of Changes Made

## 1. Case-Insensitive Title Check
- Updated the article title check to be case-insensitive using `func.lower()`.

## 2. Validation for Required Fields
- Added validation to ensure that `title`, `image_url`, and `content` fields are included in the request.
- If any are missing, an appropriate error message is returned.

## 3. Removed `used_keywords` from Request Body
- Updated the request body documentation to remove the `used_keywords` field.
- Clarified that this field is not required.

## 4. Updated `analysis` Field Documentation
- Modified the documentation to indicate that the `analysis` field can be left empty.

## 5. Default Value for `is_article_efficient`
- Updated the `is_article_efficient` field to default to `'Green - {comment or empty}'` when creating an article.

## 6. Updated URL String
- Changed the URL string to `"Generated By AI Alpha Team"`.

## 7. Detailed Required Parameters in Documentation
- Enhanced the docstring and Swagger JSON to clearly specify which parameters are required for the request.
